Add ERC: Expirable ERC-20 (#718)
ethereum/ERCs: ERCS/erc-7818.md

## Motivation

This extension facilitates the development of [ERC-20](./eip-20.md) standard compatible tokens featuring expiration dates. This capability broadens the scope of potential applications, particularly those involving time-sensitive assets. Expirable tokens are well-suited for scenarios necessitating temporary validity, including

- Bonds or financial instruments with defined maturity dates
- Time-constrained assets within gaming ecosystems
- Next-gen loyalty programs incorporating expiring rewards or points
- Prepaid credits for utilities or services (e.g., cashback, data packages, fuel, computing resources) that expire if not used within a specified time frame
- Postpaid telecom data package allocations that expire at the end of the billing cycle, motivating users to utilize their data before it resets
- Tokenized e-Money for a closed-loop ecosystem, such as transportation, food court, and retail payments

## Specification

The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> Implementers of this **REQUIRED** inheritance [ERC-20](./eip-20.md) interface and **MUST** have all of the following functions and all function behavior **MUST** meet the specification.

Reviewer: ```suggestion
Compatible implementations MUST inherit from [ERC-20](./eip-20.md)'s interface and **MUST** have all of the following functions and all function behavior **MUST** meet the specification.
```

```solidity
// SPDX-License-Identifier: CC0-1.0
pragma solidity >=0.8.0 <0.9.0;

/**
 * @title ERC-7818: Expirable ERC20
 * @dev Interface for creating expirable ERC20 tokens.
 */

> import "@openzeppelin/contracts/token/ERC20/IERC20.sol";

Reviewer: Don't use open zeppelin within your specification. You can assume implementers will find the interface from ERC-20 itself.

interface IERC7818 is IERC20 {

    /**
     * @dev Retrieves the balance of a specific `identifier` owned by an account.
     * @param account The address of the account.

>      * @param identifier The Identifier "MAY" represents an epoch, round, period, or token identifier.

Reviewer: You should probably introduce the concept of epoch/identifiers _before_ diving into the interface. Maybe add a bit more introductory text explaining how this standard is meant to work before the code fence?

     * @return uint256 The balance of the specified `identifier`.
     * @notice `identifier` "MUST" represent a unique identifier, and its meaning "SHOULD"

>      * align with how contract maintain the `epoch` in the implementing contract.

Reviewer: Putting requirements in the comments is allowed, but I find it a bit more difficult to read. You might want to move them outside the code fence? This applies to all the netspec comments.

     */
    function balanceOf(
        address account,
        uint256 identifier
    ) external view returns (uint256);

    /**
     * @dev Retrieves the current epoch of the contract.
     * @return uint256 The current epoch of the token contract,
     * often used for determining active/expired states.
     */
    function epoch() external view returns (uint256);

    /**
     * @dev Retrieves the duration a token remains valid.
     * @return uint256 The validity duration.
     * @notice `duration` "MUST" specify the token's validity period.
     * The implementing contract "SHOULD" clearly document,
     * whether the unit is blocks or time in seconds.
     */
    function duration() external view returns (uint256);

    /**
     * @dev Checks whether a specific `identifier` is expired.

>      * @param identifier The Identifier "MAY" represents an epoch, round, period, or token identifier.

Reviewer: So since you have a function `epoch`, does it make sense to also include "round" / "period" as well? Seems like the standard explicitly names it "epoch".

>      * @param identifier The Identifier "MAY" represents an epoch, round, period, or token identifier.

Reviewer: This is ERC-20 based, so what is a "token identifier"?

     * @param identifier The Identifier "MAY" represents an epoch, round, period, or token identifier.
     * @param value The amount to transfer.
     * @return bool True if the transfer succeeded, false or reverted if give `identifier` it's expired.
     * @notice The transfer "MUST" revert if the token `identifier` is expired.
     */
    function transfer(
        address to,
        uint256 identifier,
        uint256 value
    ) external returns (bool);

    /**
     * @dev Transfers a specific `identifier` and value from one account to another.
     * @param from The sender's address.
     * @param to The recipient's address.
     * @param identifier The Identifier "MAY" represents an epoch, round, period, or token identifier.
     * @param value The amount to transfer.
     * @return bool True if the transfer succeeded, false or reverted if give `identifier` it's expired.
     * @notice The transfer "MUST" revert if the token `identifier` is expired.
     */
    function transferFrom(
        address from,
        address to,
        uint256 identifier,
        uint256 value
    ) external returns (bool);
}
```

### Behavior specification

> - `balanceOf` **MUST** return the total balance of tokens held by an account that are still valid (i.e., have not expired). This includes any tokens associated with specific periods, epochs, or other identifiers, provided they remain within their validity duration. Expired tokens **MUST NOT** be included in the returned balance, ensuring that only actively usable tokens are reflected in the result.

Reviewer: This is ambiguous. Do you mean _just_ the newly introduced `balanceOf` or both this one and the original?

- `transfer` and `transferFrom` **MUST** exclusively transfer tokens that remain non-expired at the time of the transaction. Attempting to transfer expired tokens **MUST** revert the transaction or return false. Additionally, implementations **MAY** include logic to prioritize the automatic transfer of tokens closest to expiration, ensuring that the earliest expiring tokens are used first, provided they meet the non-expired condition.
- `totalSupply` **SHOULD** be set to `0` or `type(uint256).max` due to the challenges of tracking only valid (non-expired) tokens.

> - The implementation **MAY** use a standardized revert message, such as `ERC7818TransferredExpiredToken` or `ERC7818TransferredExpiredToken(address sender, uint256 identifier)`, to clearly indicate that the operation failed due to attempting to transfer expired tokens. 

Reviewer: Are these custom error types or strings?

## Rationale

> The rationale for developing an expirable [ERC-20](./eip-20.md) token extension is based on several key requirements that ensure its practicality and adaptability for various applications to

Reviewer: Incomplete sentence?

Also this belongs in motivation (since it's justifying the whole document and not a specific decision you've made within it.)

### Compatibility with the existing [ERC-20](./erc-20.md) standard. 
The extension should integrate smoothly with the [ERC-20](./eip-20.md) interface, This ensures compatibility with existing token ecosystems and third-party tools like wallets and blockchain explorers.

### Flexible interface for various implementation. 
The smart contract should be extensible, allowing businesses to tailor the expiration functionality to their specific needs like expiry in bulk or each token independent expire, whether it’s dynamic reward systems or time-sensitive applications.

> 

Reviewer: You'll definitely want to talk about why you introduced the concept of epochs in here. It doesn't seem necessarily related to expiry.

## Backwards Compatibility

This standard is fully [ERC-20](./erc-20.md) compatible.

## Reference Implementation

> For reference implementation can be found [here](../assets/erc-7818/README.md), But in the reference implementation, we employ a sorted list to automatically select the token that nearest expires first with a First-In-First-Out (`FIFO`) and sliding window algorithm that operates based on the `block.number` as opposed to relying on `block.timestamp`, which has been criticized for its lack of security and resilience, particularly given the increasing usage of Layer 2 (L2) networks over Layer 1 (L1) networks. Many L2 networks exhibit centralization and instability, which directly impacts asset integrity, rendering them potentially unusable during periods of network halting, as they are still reliant on the timestamp.

Reviewer: ```suggestion
For reference implementation can be found [here](../assets/eip-7818/README.md), But in the reference implementation, we employ a sorted list to automatically select the token that nearest expires first with a First-In-First-Out (`FIFO`) and sliding window algorithm that operates based on the `block.number` as opposed to relying on `block.timestamp`, which has been criticized for its lack of security and resilience, particularly given the increasing usage of Layer 2 (L2) networks over Layer 1 (L1) networks. Many L2 networks exhibit centralization and instability, which directly impacts asset integrity, rendering them potentially unusable during periods of network halting, as they are still reliant on the timestamp.
```

## Security Considerations

> ### SC06: Denial Of Service

Reviewer: What are these tags ("SC06")?

Author: It's from the OWASP Top 10 dApp list https://github.com/OWASP/www-project-smart-contract-top-10/blob/main/2023/en/src/SC06-denial-of-service-attacks.md but yeah I decided to remove it since the DoS and other vulnerability concept is already well and can easily be searched on GitHub or in a browser.
